pkg/network/ebpf/c/prebuilt/http.c

```diff
@@ -240,6 +255,147 @@ int uprobe__SSL_shutdown(struct pt_regs* ctx) {
+    init_ssl_sock(ssl_session, (u32)fd);
+    return 0;
+}
+
+// void gnutls_transport_set_ptr2 (gnutls_session_t session, gnutls_transport_ptr_t recv_ptr, gnutls_transport_ptr_t send_ptr)
+// "In berkeley style sockets this function will set the connection descriptor."
+SEC("uprobe/gnutls_transport_set_ptr2")
+int uprobe__gnutls_transport_set_ptr2(struct pt_regs* ctx) {
+    void *ssl_session = (void *)PT_REGS_PARM1(ctx);
+    // Use the recv_ptr and ignore the send_ptr;
+    // in most real-world scenarios, they are the same.
+    // This should include the file descriptor cast as a pointer
+    void *recv_fd_as_ptr = (void *)PT_REGS_PARM2(ctx);
+
+    int recv_fd = (int)(uintptr_t)recv_fd_as_ptr;
+    init_ssl_sock(ssl_session, (u32)recv_fd);
+    return 0;
+}
+
+// ssize_t gnutls_record_recv (gnutls_session_t session, void * data, size_t data_size)
+SEC("uprobe/gnutls_record_recv")
+int uprobe__gnutls_record_recv(struct pt_regs* ctx) {
+    void *ssl_session = (void *)PT_REGS_PARM1(ctx);
+    void *data = (void *)PT_REGS_PARM2(ctx);
+
+    // Re-use the map for SSL_read
+    ssl_read_args_t args = {
+        .ctx = ssl_session,
+        .buf = data,
+    };
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    bpf_map_update_elem(&ssl_read_args, &pid_tgid, &args, BPF_ANY);
+    return 0;
+}
+
+// ssize_t gnutls_record_recv (gnutls_session_t session, void * data, size_t data_size)
+SEC("uretprobe/gnutls_record_recv")
+int uretprobe__gnutls_record_recv(struct pt_regs* ctx) {
+    ssize_t read_len = (ssize_t)PT_REGS_RC(ctx);
+
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    // Re-use the map for SSL_read
+    ssl_read_args_t *args = bpf_map_lookup_elem(&ssl_read_args, &pid_tgid);
+    if (args == NULL) {
+        return 0;
+    }
+
+    void *ssl_session = args->ctx;
+    conn_tuple_t *t = tup_from_ssl_ctx(ssl_session, pid_tgid);
+    if (t == NULL) {
+        goto cleanup;
+    }
+
+    char buffer[HTTP_BUFFER_SIZE];
+    __builtin_memset(buffer, 0, sizeof(buffer));
+    if (read_len >= HTTP_BUFFER_SIZE) {
+        bpf_probe_read(buffer, sizeof(buffer), args->buf);
+    }
+
+    skb_info_t skb_info = {0};
+    __builtin_memcpy(&skb_info.tup, t, sizeof(conn_tuple_t));
+    http_process(buffer, &skb_info, skb_info.tup.sport);
+ cleanup:
+    bpf_map_delete_elem(&ssl_read_args, &pid_tgid);
+    return 0;
+}
+
+// ssize_t gnutls_record_send (gnutls_session_t session, const void * data, size_t data_size)
+SEC("uprobe/gnutls_record_send")
+int uprobe__gnutls_record_send(struct pt_regs* ctx) {
+    void *ssl_session = (void *)PT_REGS_PARM1(ctx);
+    void *data = (void *)PT_REGS_PARM2(ctx);
+    size_t data_size = (size_t)PT_REGS_PARM3(ctx);
+
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    conn_tuple_t *t = tup_from_ssl_ctx(ssl_session, pid_tgid);
+    if (t == NULL) {
+        return 0;
+    }
+
+    char buffer[HTTP_BUFFER_SIZE];
+    __builtin_memset(buffer, 0, sizeof(buffer));
+    if (data_size >= HTTP_BUFFER_SIZE) {
```

Comment:
Should we handle the case where `data_size < HTTP_BUFFER_SIZE`?. `GET / HTTP/1.0\r\n\r\n` is a valid HTTP request and is only 22 bytes long.

Reply:
I added some logic to handle this path. The reason I originally had this code is because this is also how the OpenSSL tracing behaves, so that will continue to miss data if shorter than the buffer size (unless it also gets changed).

Reply:
I took a quick look and I think `httpTX.Path` might need to be updated to handle null termination earlier than the length of the buffer.

Reply:
This should now be fixed, and I added a (somewhat unrealistic) unit test that tests the behavior.